nengo/utils/sockets.py

```diff
@@ -0,0 +1,322 @@
+from __future__ import absolute_import
+
+import struct
+import socket
+import time
+import threading
+import errno
+import warnings
+
+from nengo.utils.compat import queue
+
+
+class SocketCheckAliveThread(threading.Thread):
+    def __init__(self, socket_class):
+        threading.Thread.__init__(self)
+        self.socket_class = socket_class
+
+    def run(self):
+        # Keep checking if the socket class is still being used.
+        while (time.time() - self.socket_class.last_active <
+               self.socket_class.max_idle_time_current):
+            time.sleep(self.socket_class.alive_thread_sleep_time)
+        # If the socket class is idle, terminate the sockets
+        self.socket_class._close_recv_socket()
+        self.socket_class._close_send_socket()
+
+
+class UDPSocket:
```

Comment:
Documentation (in the form of docstrings) for the internal workings of this class (or at least the `run` function) would be appreciated.